go/consensus/cometbft/apps/vault/action.go

```diff
@@ -0,0 +1,76 @@
+package vault
+
+import (
+	"github.com/oasisprotocol/oasis-core/go/consensus/cometbft/api"
+	vaultState "github.com/oasisprotocol/oasis-core/go/consensus/cometbft/apps/vault/state"
+	vault "github.com/oasisprotocol/oasis-core/go/vault/api"
+)
+
+// executeAction executes a given action in the context of a vault. Assumes the action has already
+// been validated before execution.
+func (app *vaultApplication) executeAction(ctx *api.Context, vlt *vault.Vault, action *vault.Action) error {
+	switch {
+	case action.Suspend != nil:
+		// Suspend a vault.
+		oldState := vlt.State
+		vlt.State = vault.StateSuspended
+
+		ctx.EmitEvent(api.NewEventBuilder(app.Name()).TypedAttribute(&vault.StateChangedEvent{
+			Vault:    vlt.Address(),
+			OldState: oldState,
+			NewState: vlt.State,
+		}))
+	case action.Resume != nil:
+		// Resume a vault.
+		oldState := vlt.State
+		vlt.State = vault.StateActive
+
+		ctx.EmitEvent(api.NewEventBuilder(app.Name()).TypedAttribute(&vault.StateChangedEvent{
+			Vault:    vlt.Address(),
+			OldState: oldState,
+			NewState: vlt.State,
+		}))
+	case action.UpdateWithdrawPolicy != nil:
+		// Update withdraw policy for an address.
+		state := vaultState.NewMutableState(ctx.State())
+		addrState, err := state.AddressState(ctx, vlt.Address(), action.UpdateWithdrawPolicy.Address)
+		switch err {
+		case nil:
+		case vault.ErrNoSuchState:
+			addrState = &vault.AddressState{}
+		default:
+			return err
+		}
+
+		addrState.WithdrawPolicy = action.UpdateWithdrawPolicy.Policy
```

Comment:
Bucket and amount could be set to 0 after update. Or only if `LimitInterval` changes.